Clean up APIv3 error handling

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/19048
## Description

This is an early start for improvements to the error handling in APIv3. I initially wanted to include these changes in another PR, but they are not needed there, so I postponed them to **this** PR, which will hopefully be finished in the coming weeks.
## Content
- merging of model and contract errors moved into the contract
  - also removed older (prior to this PR not working) validation copy from contract, which had the same purpose
- `contract.errors` are now used like they are intended to be used:
  - first argument is a symbol for the name of the column that is errorneous (`:base` for general errors)
  - second argument is a symbol that describes the error (and gets translated by AR)
- To generate apropriate responses based on different errors, the `ActiveModel::Errors` were patched (further) to get the ability to tell the symbol used as error (instead of always returning a localized string)
- Completely reworked translation of `ActiveRecord::Errors` into `API::Errors::ErrorBase`
  - major work is done by `ErrorBase::Create`
  - creation of `multipleErrors` now not a concern of `Validation` errors anymore
  - `ActiveRecord::Errors` should now be used in fewer places, API errors in more
  - errors should now always correctly report the affected property (correct as in: They should refer to it by its API name)

NobodysNightmare committed:
extract error merging into contract

NobodysNightmare committed:
remove side effect from Validation.create

I just don't know what the original author thought...

NobodysNightmare committed:
rename local

NobodysNightmare committed:
hide error merging behind contract.validate

NobodysNightmare committed:
fix stupid mistakes

- reference to method by old name
- not knowing how to call a super-method

NobodysNightmare committed:
dirty trick to allow using a Hash as mock

also removes default value from initializer, which was never sane

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-error-cleanup.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-error-cleanup.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
style

[ci skip]

NobodysNightmare committed:
use contract directly in WP create too

NobodysNightmare committed:
resort errors in WP contracts

- use attributes as keys and error messages as value
- AM errors immediately turn message symbols into translated strings, this makes them hardly processable
-> patch some helper methods onto AM::Errors and everything is fine forever

NobodysNightmare committed:
remove tautology and add apropriate specs

NobodysNightmare committed:
adapt specs to new contract behaviour

- also refine specs for done-ratio to check for non-fail when not changed
- use shared examples for parent properties also in done ratio specs

NobodysNightmare committed:
define properties on contract

Actually it is documented that a :property needs to be defined on the contract (see https://github.com/apotonick/reform#contracts).

We just did not do that until now, because it was not neccessary when abusing AR::Errors as hard as we did.

NobodysNightmare committed:
treat all 422 errors the same for forms

That means: Do render the form for all HTTP 422 errors, problems are highlighted within the form

NobodysNightmare committed:
mock the model in model_contract_spec

NobodysNightmare committed:
fix bogus errors in contracts

- AR tried to read properties that are undefined on the contract (:error_read_only)
- now that we actually use Contract properties the copy-validations stuff started to do things, however:
    - we already went another path to simultaneously validate contract + model
    - validation was apparently broken, e.g. though the subject was not empty, there were subject :blank errors

NobodysNightmare committed:
rework error instantiation

- multiple errors are no longer a special case of specific errors
- there is a separate error class to represent multiple API errors
- convert ActiveRecord errors to API errors in a single place, ideally work with API errors from there on

NobodysNightmare committed:
include punctuation in error locales

NobodysNightmare committed:
apply basic attribute name transformations

those are the ones that were present earlier already. They do not cover the hard cases,
where the backend has completely different names form the API

NobodysNightmare committed:
apply a hack to fix an error

NobodysNightmare committed:
fix some specs

- error messages now contain punctuation
- when :save returns false, there should be errors present
- errors in form representers are now API errors

NobodysNightmare committed:
fix some errors uncovered by specs

- bad default values
- missing attribute sanitation
- style

NobodysNightmare committed:
store messages alongside error symbols

- allows to skip translation part in ErrorBase
- allows to have correct translations containing placeholder values

NobodysNightmare committed:
extract attribute name conversion into extra class

- consider special renaming rules for well known attribute names
-> currently covering wp attributes and createdAt, updatedAt

NobodysNightmare committed:
use merge! method of Reform::Contract::Errors

- this is more concise, but more importantly:
- it uses our patched in symbol merging, thereby correctly transferring symbol information

NobodysNightmare committed:
change validation of WP children

- do not add error on a symbol called :#123 (123 is the childrens ID)
    - this is against the semantics of ActiveRecord::Errors
    - creating symbols in a loop used to be a bad idea (might be better today)
    - it breaks error messages in the APIv3 (which is due to APIv3 working beyond the borders of AR errors, not the best of ideas...)
- use :base instead, the place for unlocatable errors
- include the fact that #123 refers to a child element in the error message

Basically I feel that this makes the children validation cleaner and better. But somehow I fear, that there is some code out there that depends on the fact that error keys like :#123 existed...

NobodysNightmare committed:
use common name conversion in linked properties

some day we could do that for ALL properties...

NobodysNightmare committed:
style

[ci skip]

NobodysNightmare commented:
I am somehow hoping, that tests will now pass and if they do I'd consider this PR done.

Note that I do not intend to fix the three line length warnings in specs where I solely added a `.`.
This would require me to extensively reformat existing test classes, because if I would only reformat the touched lines they would not fit the rest of their specs... I think this PR is already big enough.

NobodysNightmare committed:
use AR errors as mock

NobodysNightmare commented:
@ulferts @jonasheinrich @RobinWagner This PR contains changes to **existing** translations.

Since we moved the translations to Crowdin the german translations are now no longer part of this PR, however they still need to be applied.

How do we ensure that this is done?

Also: Timing might be critical here, because some specs [actually cover german translations](https://github.com/NobodysNightmare/openproject/commit/138d3dba298d937ddce58f2d255fb4fd68688cbc#diff-08669e2e1d139f5aa66cb28519dd531aR178).

edit: Bonus question: Can we make sure that the new translations only apply to `release/4.3` and above, while the versions below will keep their old punctuation?

NobodysNightmare commented:
Note: I expect to get a **failing** travis build here, see my comment about testing german translations above.

This does not seem to be fixable within this PR, because either this PR or the corresponding release branch will be red.

NobodysNightmare commented:
It might be sensible to remove the german expectations from the specs since german is no longer a language that is provided by the core. Thus, it should arguably also not be part of core specs...

Any opinions?

NobodysNightmare committed:
remove spec expecting german translations

german translations are not part of OpenProject anymore (they are now provided via Crowdin)

NobodysNightmare commented:
I removed this particular spec, no more german translations in specs! ^^

oliverguenther commented on config/initializers/10-patches.rb:
```diff
@@ -93,9 +125,48 @@ def full_message(attribute, message)
+      attribute = attribute.to_sym # avoid mixed symbol/string keys
+      error_symbols[attribute] = [] unless error_symbols[attribute]
+
+      error_symbols[attribute]
```
I'm all for explicit, but this seems somewhat redundant to me. Isn't the whole method body effectively `error_symbols[attribute.to_sym] ||= []` ?

NobodysNightmare replied:
I think you are right.

Admittedly I often find myself adding a last line to ensure the correct return value, even though it might not be neccessary.

However, I agree that this case is crystal clear ^^

NobodysNightmare committed:
be more concise!

oliverguenther commented on lib/api/contracts/model_contract.rb:
```diff
@@ -42,12 +42,27 @@ def self.attribute_validations
+          send attribute
+        end
+
+        nil
```
Doesn't this always return nil, even if `attribute` exists in the Model? I'm not entirely sure what this method should do, as the original is [just an alias to send](https://github.com/rails/rails/blob/fe664e8ce97b5f7a1e1d95dd4df0fbdba42c7f2b/activemodel/lib/active_model/validations.rb#L390).

``` ruby
class Bar
  def a
    '1234'
    end
  def b
    if true
      send :a
    end

    nil
    end
end
```

```
2.1.6 :069 > Bar.new.a
 => "1234"
2.1.6 :070 > Bar.new.b
 => nil
```

If it is to retrieve the value of `attribute`, I assume nil should either be in an else block, or you simply remove it, as it will return nil implicitly.

NobodysNightmare replied:
Thanks for spotting that.
I am actually not sure whether this method will even be called in a place where I "care", but I thought I probably should. And I did so by providing a flawed implementation ^^

oliverguenther commented on lib/api/errors/error_base.rb:
```diff
@@ -30,33 +30,64 @@
+        # In case the given errors contain 'critical' errors, the resulting Array will only
+        # contain the critical error and no non-critical errors (avoiding information disclosure)
+        # That means: The returned errors are always safe for display towards a user
+        def create_errors(errors)
```
Is the number of these critical errors likely to grow in the foreseeable future? The refactor here is fine for three errors, but I can see this getting out of hand If you add a new error.

**It is certainly out of scope for this PR, being as large as it is**, but when new errors are added, we could probably extract this method and simply iterate over all possibly errors to let them check if they apply.

NobodysNightmare replied:
I hope the list will not grow any further. Actually I'd even like the conflict to be non-critical in the future, making the list even smaller...

oliverguenther commented on lib/api/errors/validation.rb:
```diff
@@ -32,56 +32,11 @@ module Errors
-            # punctuation mark. Then it should be fair enough to assume that we
-            # don't need to prepend the error key if the error ends with a
-            # punctuation mark. Let's hope that this is true for the languages
-            # we'll support in OpenProject.
```
Genius. You will probably want to save this comment to tell your grandkids.

oliverguenther commented on lib/api/utilities/property_name_converter.rb:
```diff
@@ -0,0 +1,70 @@
+          updated_on: 'updatedAt',
+          remaining_hours: 'remainingTime',
+          spent_hours: 'spentTime'
+        }
```
I see this is necessary for compatibility and can't think of a better place to put this conversion, but we should strive to rid this in the next API revision.

NobodysNightmare replied:
Some of those are not really something we can "fix in the API", but we should fix in the underlying code: Most prominently `created_at` vs `created_on`.
In that specific case the API is simply more consistent than the existing code (the API **always** uses `createdAt`), however that does not mean it can't be fixed. It is just that probably nobody wants to write a database migration for the sake of consistency ^^

But I am all in to make the code more consistent and by that reduce the size of this list :)

oliverguenther commented on spec/models/work_package/work_package_planning_spec.rb:
```diff
@@ -170,36 +170,25 @@
     end
 
     describe 'parent' do
-      let (:de_message) { 'darf kein Meilenstein sein' }
```
Was there a decision regarding the general location for localized specs? I'd argue this shouldn't be a concern here anymore since your refactoring is no longer locale-dependent.

NobodysNightmare replied:
I asked the others here after realizing that our latest move to crowdin obsoletes such specs.

The reasoning was basically: We do not have german translations in the core anymore, so we don't need to test german translations (mixed with: "WTF did anyone test this?").

So I think at least since our move to crowdin there will be no more localized specs for the core.

oliverguenther replied:
With that comment block regarding AR error messages you removed earlier I can see _why_ someone decided to add localized specs, but I WTFed at the whole concept / assumption that this originated from. Either way, I'm glad they're gone :grimacing:

NobodysNightmare committed:
only return nil for _unknown_ properties
